File: legacy/trace_viz/embedded.py
"""Embedded rendering entry point for the Trace Visualizer.

When the dashboard is iframed by Ora, the URL carries `?session_id=<oraSessionId>`
and `?agent_type=<opencode|claude_code>`. Ora has already resolved the Ora session
id to a concrete trace file path and written a small **locator** JSON to the
conventional Ora app-data directory:

    <appDataDir>/dashboard/<oraSessionId>.json   ->  {"traceFilePath": "...", "agentType": "..."}

This module owns the embedded side of that contract: it computes the locator
root by OS convention (mirroring Ora's `app_data_dir`), reads the locator, reads
the trace file bytes, dispatches the matching parser, and hands the
`ParseResult` to the view's body renderer. It is deliberately free of Streamlit
side effects so the locator/parse path is unit-testable without a running app.
"""

from __future__ import annotations

import json
import os
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, Optional

from trace_viz.models import ParseResult
from trace_viz.parsers import claude_code as cc_parser
from trace_viz.parsers import opencode as oc_parser

# Mirrors the Ora desktop Tauri identifier so both sides resolve the same
# app-data directory without per-request coordination (decision: the locator
# root is a conventional path, not a URL/env parameter).
ORA_APP_IDENTIFIER = "space.ora.desktop"
LOCATOR_SUBDIR = "dashboard"


# ── Locator root ───────────────────────────────────────────────


def locator_root() -> Path:
    """Returns the conventional directory where Ora writes dashboard locators.

    Mirrors Ora's Tauri `app_data_dir` for the `space.ora.desktop` identifier so
    the dashboard and Ora agree on a path without any URL or env-var handoff.
    Only consulted when embedded (a `session_id` query param is present); the
    standalone/gemini flow never calls this.
    """
    home = Path.home()
    if sys.platform == "win32":
        base = os.environ.get("APPDATA")
        if base:
            return Path(base) / ORA_APP_IDENTIFIER / LOCATOR_SUBDIR
        # Defensive: APPDATA is effectively always set on Windows; fall back to
        # the Roaming profile under the user home if an exotic environment omits it.
        return home / "AppData" / "Roaming" / ORA_APP_IDENTIFIER / LOCATOR_SUBDIR
    if sys.platform == "darwin":
        return home / "Library" / "Application Support" / ORA_APP_IDENTIFIER / LOCATOR_SUBDIR
    # Linux / other POSIX: XDG data home, defaulting to ~/.local/share.
    xdg = os.environ.get("XDG_DATA_HOME")
    base = Path(xdg) if xdg and Path(xdg).is_absolute() else home / ".local" / "share"
    return base / ORA_APP_IDENTIFIER / LOCATOR_SUBDIR


# ── Locator + trace loading ───────────────────────────────────


@dataclass(frozen=True)
class Locator:
    """The Ora-written handoff that points the dashboard at a concrete trace file."""

    trace_file_path: str
    agent_type: str


def load_locator(session_id: str, root: Optional[Path] = None) -> Optional[Locator]:
    """Reads and decodes the locator Ora wrote for one Ora session id.

    Returns None when the locator is absent so the caller can surface a friendly
    "trace not ready" state instead of crashing the embedded iframe.
    """
    root = root if root is not None else locator_root()
    locator_path = root / f"{session_id}.json"
    if not locator_path.is_file():
        return None
    try:
        payload = json.loads(locator_path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    trace_file_path = payload.get("traceFilePath")
    agent_type = payload.get("agentType")
    if not isinstance(trace_file_path, str) or not isinstance(agent_type, str):
        return None
    return Locator(trace_file_path=trace_file_path, agent_type=agent_type)


def read_trace_bytes(locator: Locator) -> Optional[bytes]:
    """Reads the resolved trace file, returning None when it is not yet on disk."""
    path = Path(locator.trace_file_path)
    if not path.is_file():
        return None
    try:
        return path.read_bytes()
    except OSError:
        return None


# ── Parse dispatch ─────────────────────────────────────────────

# Canonical agent_type -> (parser, view module). agent_type values are the same
# strings Ora normalizes its AgentCli enum into before writing the locator, and
# match the dashboard's existing app_mode dispatch, so no extra mapping is needed.
def _parse_opencode(content: bytes) -> ParseResult:
    return oc_parser.parse(content)


def _parse_claude_code(content: bytes) -> ParseResult:
    return cc_parser.parse(content)


PARSE_DISPATCH: dict[str, Callable[[bytes], ParseResult]] = {
    "opencode": _parse_opencode,
    "claude_code": _parse_claude_code,
}


def parse_for_agent_type(content: bytes, agent_type: str) -> Optional[ParseResult]:
    """Dispatches bytes to the parser for the given canonical agent_type."""
    parser = PARSE_DISPATCH.get(agent_type)
    if parser is None:
        return None
    return parser(content)

File: backend/src/embedded.rs
//! Ora embedded-iframing contract — port of `legacy/trace_viz/embedded.py`.
//!
//! Ora resolves a session id to a trace file and writes a locator JSON:
//!     <appDataDir>/dashboard/<oraSessionId>.json
//!         -> {"traceFilePath": "...", "agentType": "..."}
//!
//! The locator root mirrors Ora's Tauri `app_data_dir` for the
//! `space.ora.desktop` identifier.

use std::path::PathBuf;

use serde::Deserialize;

use crate::models::ParseResult;
use crate::parsers;

pub const ORA_APP_IDENTIFIER: &str = "space.ora.desktop";
pub const LOCATOR_SUBDIR: &str = "dashboard";

/// Conventional directory where Ora writes dashboard locators (per-OS).
pub fn locator_root() -> PathBuf {
    let home = std::env::var("HOME").unwrap_or_default();
    if cfg!(windows) {
        if let Ok(appdata) = std::env::var("APPDATA")
            && !appdata.is_empty()
        {
            return PathBuf::from(appdata)
                .join(ORA_APP_IDENTIFIER)
                .join(LOCATOR_SUBDIR);
        }
        return PathBuf::from(home)
            .join("AppData")
            .join("Roaming")
            .join(ORA_APP_IDENTIFIER)
            .join(LOCATOR_SUBDIR);
    }
    if cfg!(target_os = "macos") {
        return PathBuf::from(home)
            .join("Library")
            .join("Application Support")
            .join(ORA_APP_IDENTIFIER)
            .join(LOCATOR_SUBDIR);
    }
    // Linux / other POSIX: XDG data home, defaulting to ~/.local/share.
    let base = std::env::var("XDG_DATA_HOME")
        .ok()
        .filter(|p| p.starts_with('/'))
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from(home).join(".local").join("share"));
    base.join(ORA_APP_IDENTIFIER).join(LOCATOR_SUBDIR)
}

#[derive(Debug, Deserialize)]
pub struct Locator {
    #[serde(rename = "traceFilePath")]
    pub trace_file_path: String,
    #[serde(rename = "agentType")]
    pub agent_type: String,
}

/// Read the locator Ora wrote for one session id; None when absent/invalid.
pub fn load_locator(session_id: &str, root: Option<&std::path::Path>) -> Option<Locator> {
    let root = root
        .map(std::path::Path::to_path_buf)
        .unwrap_or_else(locator_root);
    let locator_path = root.join(format!("{session_id}.json"));
    if !locator_path.is_file() {
        return None;
    }
    let payload: serde_json::Value =
        serde_json::from_str(&std::fs::read_to_string(&locator_path).ok()?).ok()?;
    let trace_file_path = payload.get("traceFilePath")?.as_str()?.to_string();
    let agent_type = payload.get("agentType")?.as_str()?.to_string();
    Some(Locator {
        trace_file_path,
        agent_type,
    })
}

/// Read the resolved trace file, None when not yet on disk.
pub fn read_trace_bytes(locator: &Locator) -> Option<Vec<u8>> {
    let path = PathBuf::from(&locator.trace_file_path);
    if !path.is_file() {
        return None;
    }
    std::fs::read(&path).ok()
}

/// One round trip mirroring `render_embedded` in legacy app.py.
#[derive(serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum EmbeddedStatus {
    Ok,
    LocatorMissing,
    AgentMismatch,
    TraceMissing,
    ParseEmpty,
    UnsupportedAgent,
}

#[derive(serde::Serialize)]
pub struct EmbeddedResponse {
    pub status: EmbeddedStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<ParseResult>,
    /// Human-readable message for the four error states (Chinese UI copy
    /// kept verbatim from legacy app.py).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

pub fn resolve_embedded(session_id: &str, agent_type: &str) -> EmbeddedResponse {
    resolve_embedded_with_root(session_id, agent_type, &locator_root())
}

/// Same flow against an explicit locator root (used by tests; production
/// passes `locator_root()`).
pub fn resolve_embedded_with_root(
    session_id: &str,
    agent_type: &str,
    root: &std::path::Path,
) -> EmbeddedResponse {
    let Some(locator) = load_locator(session_id, Some(root)) else {
        return EmbeddedResponse {
            status: EmbeddedStatus::LocatorMissing,
            result: None,
            message: Some(
                "定位器尚未生成。请先在 Ora 中打开该会话的 dashboard，让 Ora 解析并写入 trace 文件路径。"
                    .to_string(),
            ),
        };
    };
    if locator.agent_type != agent_type {
        return EmbeddedResponse {
            status: EmbeddedStatus::AgentMismatch,
            result: None,
            message: Some(format!(
                "agent_type 不一致：URL 为 {agent_type}，定位器为 {}。",
                locator.agent_type
            )),
        };
    }
    let Some(content) = read_trace_bytes(&locator) else {
        return EmbeddedResponse {
            status: EmbeddedStatus::TraceMissing,
            result: None,
            message: Some(
                "trace 文件尚未生成或为空——会话进行中或尚未产生事件，稍后再试。".to_string(),
            ),
        };
    };
    let Some(result) = parsers::parse_for_agent_type(&content, agent_type) else {
        return EmbeddedResponse {
            status: EmbeddedStatus::UnsupportedAgent,
            result: None,
            message: Some(format!("不支持的 agent_type：{agent_type}")),
        };
    };
    if result.raw_events.is_empty() {
        return EmbeddedResponse {
            status: EmbeddedStatus::ParseEmpty,
            result: None,
            message: Some("已读取 trace 文件，但未解析到任何事件。".to_string()),
        };
    }
    EmbeddedResponse {
        status: EmbeddedStatus::Ok,
        result: Some(result),
        message: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_root() -> std::path::PathBuf {
        let mut path = std::env::temp_dir();
        let nonce = std::process::id();
        path.push(format!("atv-locator-test-{nonce}"));
        std::fs::create_dir_all(&path).unwrap();
        path
    }

    #[test]
    fn locator_roundtrip() {
        let root = temp_root();
        let trace = root.join("trace.ndjson");
        std::fs::write(
            &trace,
            b"{\"type\":\"session.start\",\"model\":\"m\",\"sessionID\":\"s\",\"title\":\"t\"}\n",
        )
        .unwrap();
        let locator = serde_json::json!({
            "traceFilePath": trace.to_str().unwrap(),
            "agentType": "opencode"
        });
        std::fs::write(root.join("sess1.json"), locator.to_string()).unwrap();

        let resp = resolve_embedded_with_root("sess1", "opencode", &root);
        assert!(matches!(resp.status, EmbeddedStatus::Ok));
        let result = resp.result.unwrap();
        assert_eq!(result.session_info.session_id, "s");
        assert_eq!(result.raw_events.len(), 1);

        // missing locator
        let resp2 = resolve_embedded_with_root("sess2", "opencode", &root);
        assert!(matches!(resp2.status, EmbeddedStatus::LocatorMissing));
        assert!(resp2.message.unwrap().contains("定位器尚未生成"));

        // agent mismatch
        let resp3 = resolve_embedded_with_root("sess1", "claude_code", &root);
        assert!(matches!(resp3.status, EmbeddedStatus::AgentMismatch));

        // trace missing
        std::fs::remove_file(&trace).unwrap();
        let resp4 = resolve_embedded_with_root("sess1", "opencode", &root);
        assert!(matches!(resp4.status, EmbeddedStatus::TraceMissing));

        std::fs::remove_dir_all(&root).ok();
    }

    #[test]
    fn session_id_validation() {
        assert!(crate::api::valid_session_id("abc-123"));
        assert!(!crate::api::valid_session_id("../etc/passwd"));
        assert!(!crate::api::valid_session_id("a/b"));
        assert!(!crate::api::valid_session_id("a\\b"));
        assert!(!crate::api::valid_session_id(""));
    }
}
